### PR TITLE
[FIX] tool-install: Use restore action config options

### PR DESCRIPTION
**Pull request by edvilme**
Addresses https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2221478

If user has an invalid `PackageSource` on NuGet.config and attempts to install a tool with the `--ignore-failed-sources` attribute, it still outputs an error. 

`RestoreActionConfig` needs to be passed down to `NuGetPackageDownloader`

### CODE REVIEW

**Comment by edvilme** [comment]
Should also address https://github.com/dotnet/sdk/issues/36473

**Review by nagilson — COMMENTED** [review]
I tried this and it didn't seem to fix the issue for `tool install` or `tool restore`.
image

Let me know if you want to investigate together.
`dotnet tool update dotnet-sos --add-source "c:\Users\***\downloads" --global --ignore-failed-sources`

I used the failed source which you had provided.

I would also suggest you add a test like we discussed

**Comment by edvilme** [comment]
Missed adding the `RestoreActionConfig` to some other methods. It should be working fine now
image

**Comment by nagilson** [comment]
@baronfel Do we want to backport this to 8.0? I believe we should. Probably 8.0.3xx?

**Comment by baronfel** [comment]
When was the regression introduced?

File: src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs
Comment (nagilson) [inline src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs:54]: I see this was made `internal` so that it could be accessed via tests.
@dsplaisted What do you think of this pattern? I am not opposed to it, but Im not sure if we want this to be available in the assembly. Are people likely to start using this if we do that? If you don't have a strong opinion I want to ask @/rainersigwald 

If we decide against this, you could also test via a pattern like how  `WhenRunWithArchOptionItDownloadsAppHostTemplate` works, I believe.

**Review by nagilson — APPROVED** [review]
Fantastic job figuring the correct changes out and also investigating the tests :)
I have a remaining comment I'd like to hear on before we can merge this. There's also that final test you're working on. But the change itself looks great!

**Comment by nagilson** [comment]
> When was the regression introduced?

Sometime between .NET 8 RC2 and .NET 7, most likely rc2

File: test/dotnet.Tests/CommandTests/ToolInstallGlobalOrToolPathCommandTests.cs
Comment (nagilson) [inline test/dotnet.Tests/CommandTests/ToolInstallGlobalOrToolPathCommandTests.cs:998]: @dsplaisted Is there a standard feed we can use for testing an invalid source? I couldnt find one. Im not sure I like relying on this website but I dont have anything better unless we have something internal we can use for this.

**Comment by baronfel** [comment]
Ok then I'd say we should backport all the way back to 8.0.1xx so that source build gets it too.

**Comment by edvilme** [comment]
> Fantastic job figuring the correct changes out and also investigating the tests :) I have a remaining comment I'd like to hear on before we can merge this. There's also that final test you're working on. But the change itself looks great!

Thanks a lot! And thanks a lot for all the support. 
I also had the concern on the internal vs private, I am happy to look into how the mock pattern as to not expose this unnecessarily. Let me know :) 

I have also added the tests to check if the flag works correctly (aka, it does not throw when the feeds are invalid and the `--ignore-failed-sources` flag is used

**Comment by edvilme** [comment]
Maybe we can use https://api.nuget.org/v3/invalid.json ? That way we at least have some more influence/knowledge on what goes on with the urlSent from my iPhoneOn Oct 22, 2024, at 4:49 PM, Noah Gilson ***@***.***> wrote:﻿
@nagilson commented on this pull request.



In test/dotnet.Tests/CommandTests/ToolInstallGlobalOrToolPathCommandTests.cs:
> @@ -975,6 +989,16 @@ public void SetPermission(string path, string chmodArgument)
             {
             }
         }
+
+        private string _nugetConfigWithInvalidSources = @"{
+<?xml version=""1.0"" encoding=""utf-8""?>
+<configuration>
+  <packageSources>
+    <add key=""nuget"" value=""https://api.nuget.org/v3/index.json"" />
+    <add key=""invalid_source"" value=""https://invalidsource.com/v3/index.json"" />

@dsplaisted Is there a standard feed we can use for testing an invalid source? I couldnt find one. Im not sure I like relying on this website but I dont have anything better unless we have something internal we can use for this.

**Comment by dsplaisted** [comment]
> Maybe we can use https://api.nuget.org/v3/invalid.json ? That way we at least have some more influence/knowledge on what goes on with the url

Yes, that seems like a good idea so that it's a domain we control.

File: src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs
Comment (dsplaisted) [reply src/Cli/dotnet/commands/dotnet-tool/install/ToolInstallGlobalOrToolPathCommand.cs:54]: This is fine, IMHO, I'm not too worried about us accidentally starting to abuse this in our code.

**Review by nagilson — COMMENTED** [review]
All of my comments are resolved now, so this should be good to go. while the blazor wasm test failure is not related to your change, it does look like one remaining test is failing. 

I would suggest you prioritize this issue above all else because it seems to impact the most customers.

After that, I would backport your changes to release/8.0.1xx. Don't merge that PR until you get tactics approval and then approval after that to confirm the branch and release is open for changes. You can use the bot to create backports: https://github.com/dotnet/sdk/pull/43607#issuecomment-2415034428 Like how I did here. In that linked backport PR is a template to get tactics approval. Youll need to fill that out and do manual testing to verify that the fix worked on 8.0.1xx as well.

 After that, youll need to send an email to the .NET Tactics alias with the same information asking for approval and add the label 'servicing-consider.' Definitely reach out to confirm once youve gotten it all written up or if you have questions 😄

**Comment by nagilson** [comment]
> All of my comments are resolved now, so this should be good to go. while the blazor wasm test failure is not related to your change, it does look like one remaining test is failing.
> 
> I would suggest you prioritize this issue above all else because it seems to impact the most customers.
> 
> After that, I would backport your changes to release/8.0.1xx. Don't merge that PR until you get tactics approval and then approval after that to confirm the branch and release is open for changes. You can use the bot to create backports: [#43607 (comment)](https://github.com/dotnet/sdk/pull/43607#issuecomment-2415034428) Like how I did here. In that linked backport PR is a template to get tactics approval. Youll need to fill that out and do manual testing to verify that the fix worked on 8.0.1xx as well. After that, youll need to send an email to the .NET Tactics alias with the same information asking for approval and add the label 'servicing-consider.'

By the way, why are we doing all of this... this was broken starting in 8.0 and most people are going to be on 8.0 since its the latest stable and long term support version of .NET. We want them to get this fix because it's important. But because its a stable and already released build, the quality / QA check bar is a lot higher

**Comment by edvilme** [comment]
> it does look like one remaining test is failing

Hello. Yes, I pushed a fix recently and it should be passing now on the ci/cd. I am rebasing and squashing the commits to avoid having lots of little commits.

**Comment by nagilson** [comment]
image
 You can squah and merge the PR without having to rebase or do any extra work by clicking on this button 😉 

@/rainersigwald is very good about having a clean PR history and always squashing but our team is much less loose on that sort of thing. I would generally recommend to squash and merge the PR using the GitHub option. 

Force-pushing makes it harder to review the PR, because all of the existing changes get wiped away and the order of commits are lost, so it makes it so the entire PR needs to be reviewed again. This also saves you from having to wait for the pipeline to run again, and also if there is some other test that keeps failing that's not your fault, you dont have to deal with it a second/third/etc time~

**Review by nagilson — APPROVED** [review]
Let's :shipit: merge :shipit:
(and also backport) (this change is going to make a lot of people happy)

**Comment by edvilme** [comment]
Note: because of very big differences between the branches, this cannot be cherrypicked or backported automatically.

**Pull request merged**
